File: 121-best-time-to-buy-and-sell-stock/best-time-to-buy-and-sell-stock.ts
function maxProfit(prices: number[]): number {
    let maxi: number = Number.MIN_SAFE_INTEGER;

    let ans: number = 0;

    for(let i = prices.length - 1; i >= 0; i--){
        ans = Math.max(ans, maxi - prices[i]);
        maxi = Math.max(maxi, prices[i]);
    }
    return ans;
};

File: 121-best-time-to-buy-and-sell-stock/best-time-to-buy-and-sell-stock.java
class Solution {
    public int maxProfit(int[] prices) {
        int maxi = Integer.MIN_VALUE;
        int ans = Integer.MIN_VALUE;

        for(int i = prices.length - 1; i >= 0; i--){
            maxi = Math.max(maxi, prices[i]);
            ans = Math.max(ans, maxi - prices[i]);
        }

        return ans;
    }
}